### PR TITLE
Add FeaturesGenerator

### PR DESCRIPTION
**Pull request by ralphpina**
Design docs: https://docs.google.com/document/d/1XJap6OKnzAM-C4GoL1Ubs2zKRzJSlIdzcvddOOE2Z7Q/edit?usp=sharing

TL;DR:
Given the following Optimizely Features
```json
[
  {
    "key": "new_checkout_page",
    ...
  },
  {
    "key": "new_login_page",
    ...
  },
  {
    "key": "new_sign_up_page",
    ...
  }
]
```

You would produce the following API:
```kotlin
enum class Features(val key: String) {
    NEW_CHECKOUT_PAGE("new_checkout_page"),
    NEW_LOGIN_PAGE("new_login_page"),
    NEW_SIGN_UP_PAGE("new_sign_up_page")
}
```
We would also generate the following class which can be injected by the consumer:
```kotlin
class FeaturesClient(
    private val optimizely: Optimizely,
    private val userId: String
) {
    fun isFeatureEnabled(feature: Features): Boolean =
        optimizely.isFeatureEnabled(feature.key, userId)
}
```

### CODE REVIEW

File: src/main/kotlin/com/patxi/poetimizely/generator/FeaturesGenerator.kt [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+        with(
+            FileSpec.builder(packageName, "Features")
+                .addType(featuresEnumTypeSpec(features))
+                .addType(featuresClientTypeSpec(packageName))
+                .build()
+        ) {
+            return StringWriter().also { appendable: Appendable ->
+                this.writeTo(appendable)
+            }.toString()
+        }
```
Comment (patxibocos) [inline src/main/kotlin/com/patxi/poetimizely/generator/FeaturesGenerator.kt]: How about simplifying to this?
```suggestion
    fun build(features: List<Feature>): String =
        FileSpec.builder(packageName, "Features")
            .addType(featuresEnumTypeSpec(features))
            .addType(featuresClientTypeSpec(packageName))
            .build().run {
                StringWriter().also { appendable: Appendable ->
                    this.writeTo(appendable)
                }.toString()
            }
```

File: src/main/kotlin/com/patxi/poetimizely/generator/FeaturesGenerator.kt [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+ * - hyphens
+ * - underscores
+ * - max 64 chars
+ * - NO spaces
+ *
+ * The key will be used "to determine whether the feature is on or off in your code".
+ */
+private fun String.optimizelyFeatureKeyToEnumConstant() =
+    split("-", "_")
+        .joinToString("_") { it.trim().toUpperCase() }
```
Comment (patxibocos) [inline src/main/kotlin/com/patxi/poetimizely/generator/FeaturesGenerator.kt]: I think it is a good practice to type the return of functions that are not very trivial.
```suggestion
private fun String.optimizelyFeatureKeyToEnumConstant(): String =
    split("-", "_").joinToString("_") { it.trim().toUpperCase() }
```

File: src/main/kotlin/com/patxi/poetimizely/generator/FeaturesGenerator.kt [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+        // each feature is a const with the key
+        features.forEach {
+            typeSpecBuilder.addEnumConstant(
+                it.key.optimizelyFeatureKeyToEnumConstant(),
+                TypeSpec.anonymousClassBuilder()
+                    .addSuperclassConstructorParameter("%S", it.key)
+                    .build()
+            )
+        }
+    }.build()
```
Comment (patxibocos) [inline src/main/kotlin/com/patxi/poetimizely/generator/FeaturesGenerator.kt]: I like what you are doing with also to separate building the type itself from adding the enum values 🙂
Not sure if removing the with would simplify the code or not 🤔 
```kotlin
    TypeSpec.enumBuilder("Features").primaryConstructor( // with constructor(val key: String)
        FunSpec.constructorBuilder()
            .addParameter("key", String::class)
            .build()
    ).addProperty(
        PropertySpec.builder("key", String::class)
            .initializer("key")
            .build()
    ).also { typeSpecBuilder ->
        // each feature is a const with the key
        features.forEach {
            typeSpecBuilder.addEnumConstant(
                it.key.optimizelyFeatureKeyToEnumConstant(),
                TypeSpec.anonymousClassBuilder()
                    .addSuperclassConstructorParameter("%S", it.key)
                    .build()
            )
        }
    }.build()
```

File: src/test/kotlin/com/patxi/poetimizely/generator/FeaturesGeneratorTest.kt [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+            `when`("Compiling the generated code for features") {
+                val featuresCode = generator.build(features)
+                val compilationResult = KotlinCompilation().apply {
+                    sources = listOf(SourceFile.kotlin("Features.kt", featuresCode))
+                    inheritClassPath = true
+                    messageOutputStream = System.out
+                }.compile()
+                then("Generated code compiles") {
+                    compilationResult.exitCode shouldBe KotlinCompilation.ExitCode.OK
+                    val featuresEnum = compilationResult.classLoader.loadClass("$packageName.Features")
```
Comment (patxibocos) [inline src/test/kotlin/com/patxi/poetimizely/generator/FeaturesGeneratorTest.kt]: We can assert that it is an enum 🙂 
```kotlin
featuresEnum.isEnum shouldBe true
```

File: src/test/kotlin/com/patxi/poetimizely/generator/FeaturesGeneratorTest.kt [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+                    with(featuresEnum.enumConstants) {
+                        this shouldHaveSize features.size
+                        (this[0] as Enum<*>).name shouldBe "NEW_CHECKOUT_PAGE"
+                        (this[1] as Enum<*>).name shouldBe "NEW_LOGIN_PAGE"
+                        (this[2] as Enum<*>).name shouldBe "NEW_SIGN_UP_PAGE"
```
Comment (patxibocos) [inline src/test/kotlin/com/patxi/poetimizely/generator/FeaturesGeneratorTest.kt]: I'm ok with this, but we could also go with:
```kotlin
                        this.map { (it as Enum<*>).name } shouldBe listOf(
                            "NEW_CHECKOUT_PAGE",
                            "NEW_LOGIN_PAGE",
                            "NEW_SIGN_UP_PAGE"
                        )
```

File: src/main/kotlin/com/patxi/poetimizely/generator/FeaturesGenerator.kt [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+
+/**
+ * This will produce something like this.
+ *
+ * class FeaturesClient(
+ *    private val optimizely: Optimizely,
+ *    private val userId: String
+ * ) {
+ *    fun isFeatureEnabled(feature: Features): Boolean =
```
Comment (patxibocos) [inline src/main/kotlin/com/patxi/poetimizely/generator/FeaturesGenerator.kt]: Why don't we follow a similar approach as for `ExperimentsClient` just for consistency? Having a base class + interfaces?

File: src/main/kotlin/com/patxi/poetimizely/generator/FeaturesGenerator.kt (hunk 3 above)
Comment (ralphpina) [reply src/main/kotlin/com/patxi/poetimizely/generator/FeaturesGenerator.kt]: That's fair. Changing it.

File: src/test/kotlin/com/patxi/poetimizely/generator/FeaturesGeneratorTest.kt (hunk 4 above)
Comment (ralphpina) [reply src/test/kotlin/com/patxi/poetimizely/generator/FeaturesGeneratorTest.kt]: Nice, done 😁

File: src/test/kotlin/com/patxi/poetimizely/generator/FeaturesGeneratorTest.kt (hunk 5 above)
Comment (ralphpina) [reply src/test/kotlin/com/patxi/poetimizely/generator/FeaturesGeneratorTest.kt]: That's cleaner 👍.

File: src/main/kotlin/com/patxi/poetimizely/generator/FeaturesGenerator.kt (hunk 6 above)
Comment (patxibocos) [reply src/main/kotlin/com/patxi/poetimizely/generator/FeaturesGenerator.kt]: Do you mind if I open a PR on top of yours to see how does it look? I am a LOT of free time now…

Comment (ralphpina) [reply src/main/kotlin/com/patxi/poetimizely/generator/FeaturesGenerator.kt]: Yeah, definitely.

Comment (ralphpina) [reply src/main/kotlin/com/patxi/poetimizely/generator/FeaturesGenerator.kt]: I see a benefit to adding the base classes + interfaces when I was writing the tests. Are there others?

Comment (patxibocos) [reply src/main/kotlin/com/patxi/poetimizely/generator/FeaturesGenerator.kt]: Benefits I see:
- Generator class would be easier
- Test class would also be simpler
I am opening a PR soon, _stay tuned_

Comment (ralphpina) [reply src/main/kotlin/com/patxi/poetimizely/generator/FeaturesGenerator.kt]: Is there a use case where the consumers would cast to the base classes? There Is more complexity in adding the base classes + interfaces, and I didn't think I would need them.

Comment (ralphpina) [reply src/main/kotlin/com/patxi/poetimizely/generator/FeaturesGenerator.kt]: cool, I'll wait 😁. But no need to hurry.

**Pull request merged**